### PR TITLE
Fix api version

### PR DESCRIPTION
**Pull request by blaze-burgess-msf**
Updating the API version breaks everything except login. I suspect their swagger docs have changed and we'll need to update APIM.

Current error message from console:

```
OpenAI API returned an error 404: {"error":{"code":"404","message": "Resource not found"}}
[Error: OpenAI API returned an error]
```

At time of creation, this branch only has made the basic changes necessary to make the default version the latest one. It has not attempted to fix any interactions. Depending on the root cause, we may not have to make further code changes. But we'll need to know why it's failing before we merge in any case.

### CODE REVIEW

File: pages/api/chat.ts [hunk 1]
```diff
@@ -68,8 +68,9 @@ const handler = async (req: NextRequest): Promise<Response> => {
       const token: CustomJWT = await getToken({req});
       let resp;
       try {
+        const apimUrl = `${OPENAI_API_HOST}/${APIM_CHAT_ENDPONT}/deployments/${AZURE_DEPLOYMENT_ID}/chat/completions?api-version=${OPENAI_API_VERSION}`
```
Comment (blaze-burgess-msf) [inline pages/api/chat.ts:71]: I just refactored this for debugging, but it also makes the code (marginally) more readable

File: pages/api/auth/[...nextauth].ts [hunk 2]
```diff
@@ -11,7 +11,7 @@ export const authOptions = {
       tenantId: process.env.AZURE_AD_TENANT_ID || '',
       authorization: {
           params: {
-              scope: "openid User.Read User.ReadBasic.all offline_access",
+              scope: "openid User.Read offline_access",
```
Comment (blaze-burgess-msf) [inline pages/api/auth/[...nextauth].ts:14]: Was an older fix that I apparently never pushed. We've discussed this as an unnecessary permission before, though it wasn't causing the app approval auth issue that we were dealing with at the time.

File: utils/app/const.ts [hunk 3]
```diff
@@ -43,7 +43,7 @@ export const OPENAI_API_HOST_TYPE =
     process.env.OPEN_AI_HOST_TYPE || 'apim';
 
 const COMMON_CONFIGURATION: any = {
-    OPENAI_API_VERSION: "2023-03-15-preview",
+    OPENAI_API_VERSION: OPENAI_API_VERSION,
```
Comment (blaze-burgess-msf) [inline utils/app/const.ts:46]: This is really the only part that fixes the issue of the version being hard-coded. Everything else is just setting up appropriate defaults

File: pages/api/models.ts [hunk 4]
```diff
@@ -18,16 +18,28 @@ export const config = {
...
+    if (model?.capabilities) {
+      return model.capabilities.chat_completion;
+    } else {
+      return true;
+    }
+  })
       .map((model: any) => {
-        const model_name = (configData.OPENAI_API_TYPE === 'azure') ? model.model : model.id;
+        let model_passed = true;
+        const model_name = (configData.OPENAI_API_TYPE === 'azure' && model.model) ? model.model : model.id;
```
Comment (blaze-burgess-msf) [inline pages/api/models.ts:30]: They changed the data structure of models to (more closely?) match openai's output. This accounts for that, but allows some level of backwards compatibility.

File: pages/api/models.ts [hunk 5]
```diff
@@ -57,7 +69,7 @@ const handler = async (req: NextRequest): Promise<Response> => {
 
     let url = `${configData.OPENAI_API_HOST}/v1/models`;
     if (configData.OPENAI_API_TYPE === 'azure') {
-      url = `${configData.OPENAI_API_HOST}/${APIM_MANAGEMENT_ENDPONT}/deployments?api-version=${configData.OPENAI_API_VERSION}`;
+      url = `${configData.OPENAI_API_HOST}/${APIM_MANAGEMENT_ENDPONT}/models?api-version=${configData.OPENAI_API_VERSION}`;
```
Comment (blaze-burgess-msf) [inline pages/api/models.ts:72]: Only meaningful change needed for the API endpoint is here.

File: types/openai.ts [hunk 6]
```diff
@@ -9,9 +9,12 @@ export interface OpenAIModel {
...
 export enum OpenAIModelID {
   GPT_3_5 = 'gpt-3.5-turbo',
-  GPT_3_5_AZ = 'gpt-35-turbo',
-  GPT_4 = 'gpt-4',
-  GPT_4_32K = 'gpt-4-32k',
+  GPT_3_5_AZ = 'gpt-35-turbo-1106',
+  GPT_3_5_16k = 'gpt-35-turbo-16k-0613',
+  GPT_4 = 'gpt-4-0613',
+  GPT_4_32K = 'gpt-4-32k-0613',
+  GPT_4_TURBO = 'gpt-4-1106-Preview',
```
Comment (blaze-burgess-msf) [inline types/openai.ts:16]: Just updated available models while I was already here. I've tested all the ones that are available as well as gpt-4-vision-preview locally. Everything seems to be working.

**Comment by blaze-burgess-msf** [comment]
Working locally so I'm shifting to ready for review

**Comment by mitcheman** [comment]
Need to change the production infrastructure to use the api update before this is pushed. Because we do not have a multi-region HA production env this will require down time.

**Pull request merged**